### PR TITLE
feat: add `sload` op (non-bus-mapping version)

### PR DESCRIPTION
**Pull request by HAOYUatHZ**
https://github.com/appliedzkp/zkevm-specs/pull/106

Notes:
1. would like to separate sload and sstore PR. because the gas calculation and the TxRefund calculation for `sstore` are more complicated.
2. bus-mapping version: https://github.com/appliedzkp/zkevm-circuits/pull/347

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution/storage/sload.rs
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/execution/storage/sload.rs:187]: What does `result` mean? the sload succeeds or not?

Comment (HAOYUatHZ) [reply zkevm-circuits/src/evm_circuit/execution/storage/sload.rs:181]: means persist (not reverted) or not.
In python spec (test_begin_tx.py & test_sstore.py & test_sload.py) and **previous** circuit codes, it's called `result`.
but after https://github.com/appliedzkp/zkevm-circuits/pull/292/files#diff-8f5984af7a71d8fdfb2862e9655564673740a1d72df468ede0f01dbbd3ffba98 we have `is_success` & `is_persistent`.
I am actually not sure the difference between `is_success` & `is_persistent`, but I think I'd better 
raname `result` to `is_persistent` here?

File: zkevm-circuits/src/evm_circuit/execution/storage/sload.rs
Comment (HAOYUatHZ) [reply zkevm-circuits/src/evm_circuit/execution/storage/sload.rs]: fixed in https://github.com/appliedzkp/zkevm-circuits/pull/334/commits/3314f355cf957ad93a1cd70aa97b9c52d171ae19

File: zkevm-circuits/src/evm_circuit/execution/storage.rs
Comment (HAOYUatHZ) [reply zkevm-circuits/src/evm_circuit/execution/storage.rs:1]: The reason why I put it in an isolated module is that, I want to follow a simalar structure of `memory`: both `mload` and `mstore` are put in `memory.rs`.

But implementing both `sload` and `sstore` in a single `storage.rs` seems too extensive, so I further separate it into `sload.rs` and `sstore.rs`. That is, `sload.rs` and `sstore.rs` will be in `storage` mod.

If more votes on "`sload.rs` & `sstore.rs` without `storage` mod", I would like to change my orignial structure based on your comments.

Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/execution/storage.rs:1]: I prefer to just have `sload.rs` and `sstore.rs` in the `execution` folder, but open to others opinions.

File: zkevm-circuits/src/evm_circuit/execution/storage/sload.rs
Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/execution/storage/sload.rs:181]: Yes, I think `is_persistent` is definitely better. I also wonder what's the difference between `is_success` & `is_persistent`.

Comment (ed255) [reply zkevm-circuits/src/evm_circuit/execution/storage/sload.rs]: `is_success` means that the call returns successfully.

`is_persistent` means that the state changes done in the call will persist after the tx completes.  This means that `is_persistent` requires the call to be `is_success` and all parent calls to be `is_success`.

When you have a call that `is_success` but not `is_persistent` it means that a parent call will eventually fail reverting the state changes done in this call.

File: zkevm-circuits/src/evm_circuit/witness.rs
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/witness.rs:572]: If this is going to be exported as `pub`, add some docs for it please.

File: zkevm-circuits/src/state_circuit/state.rs
Comment (CPerezz) [inline zkevm-circuits/src/state_circuit/state.rs:1606]:
```suggestion
                Word::zero(),
```

File: zkevm-circuits/src/state_circuit/state.rs
Comment (CPerezz) [inline zkevm-circuits/src/state_circuit/state.rs:1621]:
```suggestion
                Word::zero(),
```

File: zkevm-circuits/src/state_circuit/state.rs
Comment (CPerezz) [inline zkevm-circuits/src/state_circuit/state.rs:1651]:
```suggestion
                Word::zero(),
```

File: zkevm-circuits/src/state_circuit/state.rs
Comment (CPerezz) [inline zkevm-circuits/src/state_circuit/state.rs:1745]:
```suggestion
                Word::zero(),
```

File: zkevm-circuits/src/state_circuit/state.rs
Comment (CPerezz) [inline zkevm-circuits/src/state_circuit/state.rs:1760]:
```suggestion
                Word::zero(),
```

File: zkevm-circuits/src/state_circuit/state.rs
Comment (CPerezz) [inline zkevm-circuits/src/state_circuit/state.rs:1774]:
```suggestion
                Word::zero(),
```

File: zkevm-circuits/src/state_circuit/state.rs
Comment (CPerezz) [inline zkevm-circuits/src/state_circuit/state.rs:1594]:
```suggestion
                Word::zero(),
```

File: zkevm-circuits/src/state_circuit/state.rs
Comment (CPerezz) [inline zkevm-circuits/src/state_circuit/state.rs:1398]:
```suggestion
                Word::zero(),
```

File: zkevm-circuits/src/state_circuit/state.rs
Comment (CPerezz) [inline zkevm-circuits/src/state_circuit/state.rs:1381]:
```suggestion
                Word::zero(),
```

File: zkevm-circuits/src/state_circuit/state.rs
Comment (CPerezz) [inline zkevm-circuits/src/state_circuit/state.rs:1367]:
```suggestion
                Word::zero(),
```

File: zkevm-circuits/src/state_circuit/state.rs
Comment (CPerezz) [inline zkevm-circuits/src/state_circuit/state.rs:1203]:
```suggestion
                Word::zero(),
```

File: zkevm-circuits/src/evm_circuit/execution/storage/sload.rs
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/execution/storage/sload.rs:194]: We might want to add Byte storage costs or something similar in like an `enum` form. So that these values are contained in a `const` or somewhere similar and if they change or whatever, we just need to change the constant to update them all.

**Review by CPerezz — COMMENTED** [review]
A couple of nits and questions.

But so far LGTM. I don't see any inconsistencies with the specs.

**Comment by HAOYUatHZ** [comment]
> I agree with @icemelon here. Unless storage module provides benefits in regards readability or organization, we should avoid it and stick to the current format.

No problem. I will fix it

**Comment by HAOYUatHZ** [comment]
@CPerezz previously we were using `Word::from(0)` in zkevm-circuits/src/state_circuit/state.rs. 
I just change all `Word::from(0)` to `Word::zero()` according to your feedback.

File: zkevm-circuits/src/evm_circuit/execution/storage.rs
Comment (HAOYUatHZ) [reply zkevm-circuits/src/evm_circuit/execution/storage.rs]: fixed  in https://github.com/appliedzkp/zkevm-circuits/pull/334/commits/1ef16ed531862b1d7faa6adc33c9018fe4d2a3f0

**Review by CPerezz — APPROVED** [review]
LGTM!

Would like a re-review from @ed255 before this gets merged.

As for the test that can't be performed now due to being blocked by  #354, please fill an issue so that we have it listed somewhere :)

Nice job @HAOYUatHZ :smile:

File: zkevm-circuits/src/evm_circuit/execution/sload.rs
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/execution/sload.rs:33]: For these three values you're using `Word`, which allocates 32 cells for each, and when you call `expr()` you get the RLC encoding of the word.
I see that in the gadget these values are only used as RLC encoded (their bytes/limbs are never used).  So maybe there's no need to use `Word` here and you can leave them as `Cell`?  For example, check the implementation of `CALLVALUE`: https://github.com/appliedzkp/zkevm-circuits/blob/5254119c8cbecbe5aa6b6df84f427e15e05a5b08/zkevm-circuits/src/evm_circuit/execution/callvalue.rs#L24

**Review by ed255 — APPROVED** [review]
LGTM!  Thanks for working on this opcode :)

I've left a suggestion in my review, feel free to take a look and update if you consider it appropriate.  Otherwise we can proceed with merging this PR!

File: zkevm-circuits/src/evm_circuit/execution/sload.rs
Comment (HAOYUatHZ) [reply zkevm-circuits/src/evm_circuit/execution/sload.rs:33]: Sounds good, I will update it!

**Pull request merged**
